services/worker/src/worker/job_runners/config/parquet_and_info.py

```diff
@@ -840,6 +858,144 @@ def __exit__(
+    with the format "urlpath with protocol" -> {"read": int, "size": int}
+
+    Since tracking is applied directly on the file from `fs.open`, reads from file-wrappers
+    like ZipFile or GZipFile are also taken into account.
+
+    Example of usage:
+
+    ```python
+    builder = load_dataset_builder("rajpurkar/squad")
+    with track_reads() as tracker:
+        builder.download_and_prepare(file_format="parquet")
+        print(tracker.files)
+    ```
+
+    The limiter is usually used with a `StreamingDownloadManager` to not have to download
+    the full dataset:
+
+    ```python
+    builder = load_dataset_builder("rajpurkar/squad")
+    dl_manager = StreamingDownloadManager(...)
+    for split_generator in builder._split_generators(dl_manager):
+        with track_reads() as tracker:
+            builder._prepare_split(split_generator=split_generator, file_format="parquet")
+    ```
+    """
+
+    def __init__(self) -> None:
+        self.files: dict[str, dict[str, int]] = {}
+        self.exit_stack = ExitStack()
+
+    def track_read(self, urlpath: str, f_read: Callable[..., ReadOutput], *args: Any, **kwargs: Any) -> ReadOutput:
+        out = f_read(*args, **kwargs)
+        self.files[urlpath]["read"] += len(out)
+        return out
+
+    def track_iter(
+        self, urlpath: str, f_iter: Callable[..., Generator[ReadOutput, None, None]]
+    ) -> Generator[ReadOutput, None, None]:
+        for out in f_iter():
+            self.files[urlpath]["read"] += len(out)
+            yield out
+
+    def __enter__(self) -> "track_reads":
+        tracker = self
+
+        def wrapped(
+            self: Union[LocalFileSystem, HTTPFileSystem, HfFileSystem],
+            urlpath: str,
+            mode: str = "rb",
+            *args: Any,
+            fs_open: Callable[..., FsspecFile],
+            **kwargs: Any,
+        ) -> FsspecFile:
+            f = fs_open(self, urlpath, mode, *args, **kwargs)
+            urlpath = self.unstrip_protocol(urlpath)
+            if "w" not in mode:
+                f.read = functools.partial(tracker.track_read, urlpath, f.read)
+                f.__iter__ = functools.partial(tracker.track_iter, urlpath, f.__iter__)
+                if hasattr(f, "read1"):
+                    f.read1 = functools.partial(tracker.track_read, urlpath, f.read1)
+                if hasattr(f, "readline"):
+                    f.readline = functools.partial(tracker.track_read, urlpath, f.readline)
+                if hasattr(f, "readlines"):
+                    f.readlines = functools.partial(tracker.track_read, urlpath, f.readlines)
+                tracker.files[urlpath] = {"read": 0, "size": int(f.size)}
+            return f
+
+        # track files reads from local, http and hf file-systems
+        self.exit_stack.enter_context(
+            patch.object(LocalFileSystem, "open", autospec=True)
+        ).side_effect = functools.partial(wrapped, fs_open=LocalFileSystem.open)
```

Comment:
wokay I'm totally lost. I trust you

Reply:
I'll add some comments to make it clearer what it does ! (basically it replaces LocalFileSystem.open by a wrapper temporarily - the replacement is undone once the track_reads context exits)